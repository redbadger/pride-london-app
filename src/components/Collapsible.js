// @flow
import React, { PureComponent } from "react";
import { View, StyleSheet, Animated } from "react-native";
import LinearGradient from "react-native-linear-gradient";

import type { Node } from "react";
import type { LayoutEvent } from "react-native/Libraries/Types/CoreEventTypes";

import { whiteZeroColor, whiteColor } from "../constants/colors";
import text from "../constants/text";

import Touchable from "../components/Touchable";
import TextLink from "./TextLink";

type Props = {
  children: Node,
  maxHeight: number,
  showMoreLabel: string,
  showLessLabel: string
};

type State = {
  collapsed: boolean,
  contentHeight: number,
  contentContainerHeight: Object,
  gradientOpacity: Object
};

class Collapsible extends PureComponent<Props, State> {
  static defaultProps = {
    maxHeight: 140,
    showMoreLabel: text.collapsibleShowMore,
    showLessLabel: text.collapsibleShowLess
  };

  state = {
    collapsed: true,
    contentHeight: 0,
    contentContainerHeight: new Animated.Value(0),
    gradientOpacity: new Animated.Value(1)
  };

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (prevState.collapsed === this.state.collapsed) return;

    const toValue = this.state.collapsed ? 0 : 1;
    const { contentContainerHeight, gradientOpacity } = this.state;

    Animated.timing(contentContainerHeight, {
      toValue,
      duration: this.state.contentHeight * 0.8
    }).start();

    Animated.timing(gradientOpacity, {
      toValue: 1 - toValue,
      duration: this.state.contentHeight * 0.8,
      useNativeDriver: true
    }).start();
  }

  toggleCollapsed = () => {
    this.setState({ collapsed: !this.state.collapsed });
  };

  measureContentHeight = (event: LayoutEvent) => {
    this.setState({ contentHeight: event.nativeEvent.layout.height });
  };

  render() {
    const { children, showMoreLabel, maxHeight, showLessLabel } = this.props;
    const { contentHeight, collapsed } = this.state;
    const collapsible = contentHeight > maxHeight;

    const { contentContainerHeight, gradientOpacity } = this.state;
    const textContainerHeightStyle = {
      height: contentContainerHeight.interpolate({
        inputRange: [0, 1],
        outputRange: [collapsible ? maxHeight : contentHeight, contentHeight]
      })
    };
    const gradientOpacityStyle = {
      opacity: gradientOpacity.interpolate({
        inputRange: [0, 1],
        outputRange: [0, 1]
      })
    };

    return (
      <View>
        <View>
          <Animated.ScrollView
            scrollEnabled={false}
            style={textContainerHeightStyle}
          >
            <View testID="children" onLayout={this.measureContentHeight}>
              {children}
            </View>
          </Animated.ScrollView>
          {collapsible && (
            <Animated.View style={[styles.gradient, gradientOpacityStyle]}>
              <LinearGradient
                colors={[whiteZeroColor, whiteColor]}
                style={styles.gradient}
              />
            </Animated.View>
          )}
        </View>
        {collapsible && (
          <Touchable
            onPress={this.toggleCollapsed}
            style={styles.toggleContainer}
          >
            <TextLink testID="showmore">
              {collapsed ? showMoreLabel : showLessLabel}
            </TextLink>
          </Touchable>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  gradient: {
    width: "100%",
    height: 30,
    position: "absolute",
    bottom: 0
  },
  toggleContainer: {
    alignSelf: "flex-end"
  }
});

export default Collapsible;
