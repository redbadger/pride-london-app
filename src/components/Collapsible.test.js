// @flow
import React from "react";
import { Text } from "react-native";
import { shallow } from "enzyme";
import Collapsible from "./Collapsible";

it("renders correctly", () => {
  const output = shallow(
    <Collapsible>
      <Text>Hello</Text>
    </Collapsible>
  );

  expect(output).toMatchSnapshot();
});

describe("collapsing behaviour", () => {
  const render = contentHeight => {
    const output = shallow(
      <Collapsible maxHeight={200} showMoreLabel="more" showLessLabel="less">
        <Text>Hello</Text>
      </Collapsible>
    );

    const layoutEvent = output.find({ testID: "children" }).prop("onLayout");
    layoutEvent({ nativeEvent: { layout: { height: contentHeight } } });

    return output.update();
  };

  const renderAndExpand = contentHeight => {
    const collapsed = render(contentHeight);

    collapsed
      .find({ testID: "showmore" })
      .parent()
      .simulate("press");

    return collapsed.update();
  };

  describe("when content does not exceed limit", () => {
    const output = render(100);

    it("renders correctly", () => {
      expect(output).toMatchSnapshot();
    });
  });

  describe("when content exceeds limit", () => {
    const output = render(500);

    it("renders correctly", () => {
      expect(output).toMatchSnapshot();
    });

    it("shows the show more link", () => {
      expect(output.find({ testID: "showmore" }).exists()).toBe(true);
      expect(
        output
          .find({ testID: "showmore" })
          .children()
          .text()
      ).toBe("more");
    });

    it("limits the height to maxHeight");

    it("sets the gradient opacity to 1");
  });

  describe("when expanded", () => {
    const output = renderAndExpand(500);

    it("renders correctly", () => {
      expect(output.state("collapsed")).toBe(false);
      expect(output).toMatchSnapshot();
    });

    it("shows the show less link", () => {
      expect(output.find({ testID: "showmore" }).exists()).toBe(true);
      expect(
        output
          .find({ testID: "showmore" })
          .children()
          .text()
      ).toBe("less");
    });
  });
});
