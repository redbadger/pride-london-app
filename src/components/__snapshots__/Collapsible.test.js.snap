// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapsing behaviour when content does not exceed limit renders correctly 1`] = `
<Component>
  <Component>
    <AnimatedComponent
      scrollEnabled={false}
      style={
        Object {
          "height": 100,
        }
      }
    >
      <Component
        onLayout={[Function]}
        testID="children"
      >
        <Component>
          Hello
        </Component>
      </Component>
    </AnimatedComponent>
  </Component>
</Component>
`;

exports[`collapsing behaviour when content exceeds limit renders correctly 1`] = `
<Component>
  <Component>
    <AnimatedComponent
      scrollEnabled={false}
      style={
        Object {
          "height": 200,
        }
      }
    >
      <Component
        onLayout={[Function]}
        testID="children"
      >
        <Component>
          Hello
        </Component>
      </Component>
    </AnimatedComponent>
    <AnimatedComponent
      style={
        Array [
          Object {
            "bottom": 0,
            "height": 30,
            "position": "absolute",
            "width": "100%",
          },
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <LinearGradient
        colors={
          Array [
            "rgba(255, 255, 255, 0)",
            "#ffffff",
          ]
        }
        style={
          Object {
            "bottom": 0,
            "height": 30,
            "position": "absolute",
            "width": "100%",
          }
        }
      />
    </AnimatedComponent>
  </Component>
  <Touchable
    accessibilityComponentType="button"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    delayPressIn={50}
    onPress={[Function]}
    style={
      Object {
        "alignSelf": "flex-end",
      }
    }
  >
    <TextLink
      testID="showmore"
    >
      more
    </TextLink>
  </Touchable>
</Component>
`;

exports[`collapsing behaviour when expanded renders correctly 1`] = `
<Component>
  <Component>
    <AnimatedComponent
      scrollEnabled={false}
      style={
        Object {
          "height": 200,
        }
      }
    >
      <Component
        onLayout={[Function]}
        testID="children"
      >
        <Component>
          Hello
        </Component>
      </Component>
    </AnimatedComponent>
    <AnimatedComponent
      style={
        Array [
          Object {
            "bottom": 0,
            "height": 30,
            "position": "absolute",
            "width": "100%",
          },
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <LinearGradient
        colors={
          Array [
            "rgba(255, 255, 255, 0)",
            "#ffffff",
          ]
        }
        style={
          Object {
            "bottom": 0,
            "height": 30,
            "position": "absolute",
            "width": "100%",
          }
        }
      />
    </AnimatedComponent>
  </Component>
  <Touchable
    accessibilityComponentType="button"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    delayPressIn={50}
    onPress={[Function]}
    style={
      Object {
        "alignSelf": "flex-end",
      }
    }
  >
    <TextLink
      testID="showmore"
    >
      less
    </TextLink>
  </Touchable>
</Component>
`;

exports[`renders correctly 1`] = `
<Component>
  <Component>
    <AnimatedComponent
      scrollEnabled={false}
      style={
        Object {
          "height": 0,
        }
      }
    >
      <Component
        onLayout={[Function]}
        testID="children"
      >
        <Component>
          Hello
        </Component>
      </Component>
    </AnimatedComponent>
  </Component>
</Component>
`;
